// ngx-stars/esm2015/lib/ngx-stars.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class NgxStarsComponent {
    constructor() {
        this.maxStars = 5;
        this.initialStars = 0;
        this.animationSpeed = 100;
        this.wholeStars = false;
        this.ratingOutput = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.editableStars = Array.from(new Array(this.maxStars)).map((elem, index) => new EditableStar(index));
        this.setRating(this.initialStars);
        if (this.animation) {
            this.animationInterval = setInterval(this.starAnimation.bind(this), this.animationSpeed);
        }
    }
    /**
     * @return {?}
     */
    starColorAndPadding() {
        return Object.assign({}, this.starColor(), this.starPadding());
    }
    /**
     * @return {?}
     */
    starColor() {
        return { color: this.color || 'crimson' };
    }
    /**
     * @return {?}
     */
    starPadding() {
        /** @type {?} */
        let padding = '0.5rem';
        if (Number.isInteger(this.size) || this.size > 0 || this.size < 6) {
            padding = `0.${this.size}rem`;
        }
        return { 'margin-right': this.customPadding || padding };
    }
    /**
     * @return {?}
     */
    starSize() {
        if (!Number.isInteger(this.size) || this.size < 2 || this.size > 5) {
            return '';
        }
        return `fa-${this.size}x`;
    }
    /**
     * @return {?}
     */
    starAnimation() {
        this.animationRunning = true;
        if (this.rating < this.maxStars) {
            this.setRating(this.rating += 0.5);
        }
        else {
            this.setRating(0);
        }
    }
    /**
     * @return {?}
     */
    cancelStarAnimation() {
        if (this.animationRunning) {
            clearInterval(this.animationInterval);
            this.rating = 0;
            this.animationRunning = false;
        }
    }
    /**
     * @param {?} rating
     * @return {?}
     */
    setRating(rating) {
        this.rating = Math.round(rating * 2) / 2;
        this.onStarsUnhover();
    }
    /**
     * @param {?} event
     * @param {?} clickedStar
     * @return {?}
     */
    onStarHover(event, clickedStar) {
        this.cancelStarAnimation();
        /** @type {?} */
        const clickedInFirstHalf = this.clickedInFirstHalf(event);
        // fill in either a half or whole star depending on where user clicked
        clickedStar.classname = (!this.wholeStars && clickedInFirstHalf) ? 'fa-star-half-o' : 'fa-star';
        // fill in all stars in previous positions and clear all in later ones
        this.editableStars.forEach(star => {
            if (star.position > clickedStar.position) {
                star.classname = 'fa-star-o';
            }
            else if (star.position < clickedStar.position) {
                star.classname = 'fa-star';
            }
        });
    }
    /**
     * @param {?} event
     * @param {?} clickedStar
     * @return {?}
     */
    onStarClick(event, clickedStar) {
        this.cancelStarAnimation();
        /** @type {?} */
        const clickedInFirstHalf = this.clickedInFirstHalf(event);
        this.rating = clickedStar.position + ((!this.wholeStars && clickedInFirstHalf) ? 0.5 : 1);
        this.ratingOutput.emit(this.rating);
    }
    /**
     * @return {?}
     */
    onZeroStarClick() {
        this.setRating(0);
        this.ratingOutput.emit(this.rating);
    }
    /**
     * @return {?}
     */
    onZeroStarHover() {
        // clear all stars
        this.editableStars.forEach(star => star.classname = 'fa-star-o');
    }
    /**
     * @return {?}
     */
    onStarsUnhover() {
        // when user stops hovering we want to make stars reflect the last rating applied by clicking
        this.editableStars.forEach(star => {
            /** @type {?} */
            const starNumber = star.position + 1;
            if (this.rating >= starNumber) {
                star.classname = 'fa-star';
            }
            else if (this.rating > starNumber - 1 && this.rating < starNumber) {
                star.classname = 'fa-star-half-o';
            }
            else {
                star.classname = 'fa-star-o';
            }
        });
    }
    /**
     * @param {?} event
     * @return {?}
     */
    clickedInFirstHalf(event) {
        /** @type {?} */
        const starIcon = /** @type {?} */ (event.target);
        return event.pageX < starIcon.getBoundingClientRect().left + starIcon.offsetWidth / 2;
    }
    /**
     * @return {?}
     */
    noop() { }
}
NgxStarsComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-stars',
                template: `<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">

<div class="stars-line" (mouseleave)="readonly ? noop() : onStarsUnhover()">
  <i [ngClass]="['fa fa-star ' + starSize()]" class="zero-star" aria-hidden="true" (click)="onZeroStarClick()" (mousemove)="readonly ? noop() : onZeroStarHover()"></i>
  <div *ngFor="let star of editableStars;" (click)="readonly ? noop() : onStarClick($event, star)" (mousemove)="readonly ? noop() : onStarHover($event, star)">
    <i [ngClass]="['fa ' + starSize(), star.classname]" [ngStyle]="starColorAndPadding()" aria-hidden="true"></i>
  </div>
</div>
`,
                styles: [`i:not(.zero-star){color:#b42025}.stars-line{display:flex;align-items:center;position:relative}.stars-line>div{z-index:999}.zero-star{color:transparent;position:absolute;left:-16px}`],
            },] },
];
NgxStarsComponent.propDecorators = {
    maxStars: [{ type: Input }],
    initialStars: [{ type: Input }],
    readonly: [{ type: Input }],
    size: [{ type: Input }],
    color: [{ type: Input }],
    animation: [{ type: Input }],
    animationSpeed: [{ type: Input }],
    customPadding: [{ type: Input }],
    wholeStars: [{ type: Input }],
    ratingOutput: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    NgxStarsComponent.prototype.maxStars;
    /** @type {?} */
    NgxStarsComponent.prototype.initialStars;
    /** @type {?} */
    NgxStarsComponent.prototype.readonly;
    /** @type {?} */
    NgxStarsComponent.prototype.size;
    /** @type {?} */
    NgxStarsComponent.prototype.color;
    /** @type {?} */
    NgxStarsComponent.prototype.animation;
    /** @type {?} */
    NgxStarsComponent.prototype.animationSpeed;
    /** @type {?} */
    NgxStarsComponent.prototype.customPadding;
    /** @type {?} */
    NgxStarsComponent.prototype.wholeStars;
    /** @type {?} */
    NgxStarsComponent.prototype.ratingOutput;
    /** @type {?} */
    NgxStarsComponent.prototype.rating;
    /** @type {?} */
    NgxStarsComponent.prototype.editableStars;
    /** @type {?} */
    NgxStarsComponent.prototype.animationInterval;
    /** @type {?} */
    NgxStarsComponent.prototype.animationRunning;
}
export class EditableStar {
    /**
     * @param {?} position
     */
    constructor(position) {
        this.position = position;
        this.classname = 'fa-star-o';
    }
}
if (false) {
    /** @type {?} */
    EditableStar.prototype.position;
    /** @type {?} */
    EditableStar.prototype.classname;
}
